### PR TITLE
Bug 1468351 - Pre-filter client pings before partitioning in Longitudinal

### PR DESCRIPTION
**Pull request by fbertsch**
This is currently running, I expect we'll know tomorrow whether setting the value at 5000 is small enough.

### CODE REVIEW

**Comment by fbertsch** [comment]
@sunahsuh the idea here is to gather a list of allowed `(client_id, submission_date, subsession_counter)` tuples, distribute those to the executors (before we repartition the messages), and filter there.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 1]
```diff
@@ -167,25 +186,95 @@ object LongitudinalView {
...
+
+            currentClientKeysCount += 1
+
+            if(currentClientKeysCount == filterLimit){
+              val clientKeys: (String, Set[(String, Int)]) = (clientId.asInstanceOf[String], currentClientKeys.toSet)
+              allowedClientKeys += clientKeys
+            }
+          }
+
+          allowedClientKeys.iterator
```
Comment (fbertsch) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: Each partition will return a list of clients and their set of allowed `(submission_date, subsession_counter)` pairs. We can then use those to pre-filter the data in line 295.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 2]
```diff
@@ -167,25 +186,95 @@ object LongitudinalView {
...
-    deletePrefix(outputBucket, prefix)
+    s3Handler.deletePrefix(outputBucket, prefix)
 
     try {
       // Sort submissions in descending order
       implicit val ordering = Ordering[(String, String, Int)].reverse
+
+      val allowedLargeClientKeys = messages
+        .flatMap(getOrderKey)
+        .repartitionAndSortWithinPartitions(new ClientIdPartitioner(numPartitions))
```
Comment (fbertsch) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: I used `ClientIdPartitioner` and `mapPartitions` to mirror the partitioning and sorting already done, which should avoid unintended consequences. (I had considered using Dataframes, but I can't verify the behavior would be the same).

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 3]
```diff
@@ -167,25 +186,95 @@ object LongitudinalView {
...
+              allowedClientKeys += clientKeys
+            }
+          }
+
+          allowedClientKeys.iterator
+        }
+        .collect()
+        .toMap
+
+      val clientLookupMap = messages.sparkContext.broadcast(allowedLargeClientKeys)
```
Comment (fbertsch) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: Clients are only included in the `clientLookupMap` if they have more than `DefaultFilterLimit` pings. Clients who are not in the `clientLookupMap` will have all of their pings aggregated.

**Comment by fbertsch** [comment]
Good news, the run completed. Some stats:
- Took about 30 minutes longer (time spent creating the pre-filter client map)
- Only 52 clients ignored
- 1248 parquet files from a 39-node cluster (so every partition was successfully written)

This definitely fixed the issue, and 5000 is a reasonable number to set the initial value.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 4]
```diff
@@ -22,7 +22,14 @@ import scala.collection.mutable.ListBuffer
...
 
-protected class ClientIterator(it: Iterator[(String, Map[String, Any])], maxHistorySize: Int = 1000) extends Iterator[List[Map[String, Any]]] {
+import scala.collection.mutable.ListBuffer
+
+class S3Handler extends Serializable {
+  def uploadFile(localFile: java.io.File, outputBucket: String, prefix: String): Unit = S3Store.uploadFile(localFile, outputBucket, prefix)
+  def deletePrefix(bucket: String, prefix: String): Unit = utilsDeletePrefix(bucket, prefix)
+}
+
+protected class ClientIterator(it: Iterator[(String, Map[String, Any])], maxHistorySize: Int) extends Iterator[List[Map[String, Any]]] {
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: If we're removing the default of 1000 records, we should probably update the comment on the next line.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 5]
```diff
@@ -84,6 +91,18 @@ object LongitudinalView {
   // Clients are ignored when they have any bad data of some sort
   val MaxFractionIgnoredClients = .005
 
+  // This value is the total number of pings that a client must have
+  // for their pingset to be early filtered (e.g. before partitioning)
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: Please expand on "early filtered" - specifically, mentioning that clients with more pings will be excluded from the final dataset.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 6]
```diff
@@ -22,7 +22,14 @@ import scala.collection.mutable.ListBuffer
 import scala.math.abs
 import scala.reflect.ClassTag
 
-protected class ClientIterator(it: Iterator[(String, Map[String, Any])], maxHistorySize: Int = 1000) extends Iterator[List[Map[String, Any]]] {
+import scala.collection.mutable.ListBuffer
+
+class S3Handler extends Serializable {
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: Please add a comment what this class is for.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala (hunk 4 above)
Comment (fbertsch) [reply src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: Default is still 1000 records.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 7]
```diff
@@ -167,25 +200,95 @@ object LongitudinalView {
...
+    // Only show parent scalars
+    val scalarDefinitions = Scalars.definitions(includeOptin).filter(_._2.process == Some("parent"))
+
+    run(opts.to(), opts.outputBucket(), messages, sc.defaultParallelism, handler, histogramDefinitions, scalarDefinitions, DefaultFilterLimit)
+
+    sc.stop()
+  }
+
+  private def getOrderKey(message: Message): Option[((String, String, Int), Int)] = {
+    // This replicate the existing code, except returns a small data structure rather
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: nit: please specify which existing code (the `flatMap` stage of `clientMessages` processing)

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 8]
```diff
@@ -167,25 +200,95 @@ object LongitudinalView {
...
+    // than the entire ping, and doesn't filter in the for loop
+
+    val fields = message.fieldsAsMap
+
+    val payload = message.payload.getOrElse(fields.getOrElse("submission", "{}")) match {
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: `payload` is unused, this parsing can be removed. (Aside: why didn't scalastyle notice that?)

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 9]
```diff
@@ -167,25 +200,95 @@ object LongitudinalView {
...
+
+    // Only show parent scalars
+    val scalarDefinitions = Scalars.definitions(includeOptin).filter(_._2.process == Some("parent"))
+
+    run(opts.to(), opts.outputBucket(), messages, sc.defaultParallelism, handler, histogramDefinitions, scalarDefinitions, DefaultFilterLimit)
+
+    sc.stop()
+  }
+
+  private def getOrderKey(message: Message): Option[((String, String, Int), Int)] = {
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: Optional clean-up: consider introducing a separate type for this `(String, String, Int)` tuple to make it clearer that it is `clientId`, `startDate`, `profileSubsessionCounter`.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 10]
```diff
@@ -167,25 +200,95 @@ object LongitudinalView {
...
+          var currentClientKeys: ListBuffer[(String, Int)] = new ListBuffer()
+          var currentClientKeysCount: Int = 0
+
+          while(it.hasNext) {
+            val ((clientId, startDate, counter), _) = it.next()
+
+            if(clientId != currentClientId) {
+              currentClientKeys.clear()
+              currentClientKeysCount = 0
+              currentClientId = clientId.asInstanceOf[String]
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: `clientId` must already be a `String` here, the type coercion should not be needed.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 11]
```diff
@@ -167,25 +200,95 @@ object LongitudinalView {
...
+            }
+
+            if(currentClientKeysCount < DefaultMaxHistorySize) {
+              val key: (String, Int) = (startDate.asInstanceOf[String], counter.asInstanceOf[Int])
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: Same re: type coercions - the iterator is already typed.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 12]
```diff
@@ -167,25 +200,95 @@ object LongitudinalView {
...
+              currentClientKeys += key
+            }
+
+            currentClientKeysCount += 1
+
+            if(currentClientKeysCount == filterLimit){
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: style nit: add a space between `)` and `{`

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 13]
```diff
@@ -255,7 +356,8 @@ object LongitudinalView {
     } catch {
       // Delete incomplete data
       case e: Exception =>
-        deletePrefix(outputBucket, prefix)
+        e.printStackTrace()
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: Please use `logger` to print this (and ideally a message about where it blew up we can grep for)

File: src/test/scala/com/mozilla/telemetry/views/LongitudinalViewTest.scala [hunk 14]
```diff
@@ -4,28 +4,51 @@
...
+
 import scala.collection.JavaConversions._
 import scala.collection.mutable.WrappedArray
 import scala.language.reflectiveCalls
 import scala.io.Source
 
+// ScalaMock objects and functions are not serializable
+// so we'll need to mock the class ourselves
+// ref: https://groups.google.com/forum/#!topic/scalatest-users/7t7MVNzGUTQ
+class MockHandler(outputPath: String) extends S3Handler with Serializable {
```
Comment (mreid-moz) [inline src/test/scala/com/mozilla/telemetry/views/LongitudinalViewTest.scala]: Sorry for your luck :-/

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala [hunk 15]
```diff
@@ -201,20 +304,18 @@ object LongitudinalView {
               info = parse(json)
               JString(startDate) <- (info \ "subsessionStartDate").toOption
               JInt(counter) <- (info \ "profileSubsessionCounter").toOption
-            } yield ((clientId, startDate, counter.toInt),
+              subsessionCounter = counter.toInt
+
+              if(!clientLookupMap.value.contains(clientId) || clientLookupMap.value(clientId).contains((startDate, subsessionCounter)))
```
Comment (mreid-moz) [inline src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: TIL about this pattern in a `for` block. Cool!

File: src/test/scala/com/mozilla/telemetry/views/LongitudinalViewTest.scala [hunk 16]
```diff
@@ -592,15 +653,15 @@ class LongitudinalViewTest extends FlatSpec with Matchers with PrivateMethodTest
   "ClientIterator" should "not trim histories of size < 1000" in {
     val template = ("foo", Map("client" -> "foo"))
     val history = List.fill(ParentConstant)(template)
-    val split_history = new ClientIterator(history.iterator).toList
+    val split_history = new ClientIterator(history.iterator, LongitudinalView.DefaultMaxHistorySize).toList
```
Comment (mreid-moz) [inline src/test/scala/com/mozilla/telemetry/views/LongitudinalViewTest.scala]: Should these use a literal `1000` to match the test case names and list-filling and assert behaviour?

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala (hunk 9 above)
Comment (fbertsch) [reply src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: I also wanted make this a case class, but ordering of those isn't completely clear. For example, see the complications here: https://meta.plasm.us/posts/2013/10/13/ordering-case-classes/

My goal was to leave the existing code unchanged (except for adding that `if` statement) and mirror it exactly in partitioning and ordering.

File: src/test/scala/com/mozilla/telemetry/views/LongitudinalViewTest.scala (hunk 16 above)
Comment (fbertsch) [reply src/test/scala/com/mozilla/telemetry/views/LongitudinalViewTest.scala]: I'd rather keep these as-is, that way if we change that default for some reason, we have to change the tests too in order to verify the new value will work as expected.

File: src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala (hunk 9 above)
Comment (mreid-moz) [reply src/main/scala/com/mozilla/telemetry/views/Longitudinal.scala]: Ok, it makes sense to minimize the change that's being made here.

**Pull request merged**